### PR TITLE
customcluster: refactor to reduce file size of controller.go

### PR DESCRIPTION
**Pull request by Xieql**
**What type of PR is this?**

refactor to reduce file size of customcluster_controller.go and optimize code

**What this PR does / why we need it**:
current controller has exceeded 1500 Lines

**Which issue(s) this PR fixes**:
related to #209
**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
@zirain ptal

File: pkg/controllers/customcluster_controller.go [hunk 1]
```diff
@@ -201,12 +195,27 @@ func (r *CustomClusterController) Reconcile(ctx context.Context, req ctrl.Reques
...
+		Name:      cluster.Spec.ControlPlaneRef.Name,
+	}
+	kcp := &controlplanev1.KubeadmControlPlane{}
+	if err := r.Client.Get(ctx, kcpKey, kcp); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("kcp does not exist", "kcp", kcpKey)
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "failed to get kcp", "kcp", kcpKey)
+		return ctrl.Result{}, err
```
Comment (zirain) [inline pkg/controllers/customcluster_controller.go:213]: if there's a error(mostly like a network issue), should add RequeueAfter?

File: pkg/controllers/customcluster_controller.go [hunk 2]
```diff
@@ -228,719 +237,243 @@ func (r *CustomClusterController) Reconcile(ctx context.Context, req ctrl.Reques
...
 	}
 
-	if initWorker.Status.Phase == corev1.PodSucceeded {
-		customCluster.Status.Phase = v1alpha1.ProvisionedPhase
-		conditions.MarkTrue(customCluster, v1alpha1.ReadyCondition)
-		log.Info("customCluster's phase changes from Provisioning to Provisioned")
-		return ctrl.Result{}, nil
+	// get desiredClusterInfo and provisionedClusterInfo to determine if further scaling or upgrading is needed.
+	desiredClusterInfo := getDesiredClusterInfo(customMachine, kcp)
+	provisionedClusterInfo, err1 := r.getProvisionedClusterInfo(ctx, customCluster)
```
Comment (zirain) [inline pkg/controllers/customcluster_controller.go:268]: bad naming of `err1`

File: pkg/controllers/customcluster_controller.go [hunk 3]
```diff
@@ -228,719 +237,243 @@ func (r *CustomClusterController) Reconcile(ctx context.Context, req ctrl.Reques
...
+		log.Info("terminating worker was completed successfully, delete the related CRD")
+		if err := r.deleteResource(ctx, customCluster, customMachine, kcp); err != nil {
+			log.Error(err, "failed to delete resource", "customCluster", customCluster.Name)
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
 	}
-
 	if terminateWorker.Status.Phase == corev1.PodFailed {
+		log.Info(fmt.Sprintf("customCluster's phase changes from %s to %s", string(customCluster.Status.Phase), string(v1alpha1.UnknownPhase)))
```
Comment (zirain) [inline pkg/controllers/customcluster_controller.go:371]:
```suggestion
		log.Info("phase changes", "prevPhase", customCluster.Status.Phase, "currentPhase" ,v1alpha1.UnknownPhase)
```

File: pkg/controllers/customcluster_controller.go [hunk 4]
```diff
@@ -228,719 +237,243 @@ func (r *CustomClusterController) Reconcile(ctx context.Context, req ctrl.Reques
...
 	}
 
+	// After k8s cluster has been deleted successful, we need delete the related CRD.
 	if terminateWorker.Status.Phase == corev1.PodSucceeded {
-		log.Info("terminating worker was completed successfully, then we need delete the related CRD")
-		// After k8s cluster on VMs has been reset successful, we need delete the related CRD.
-		return r.reconcileDeleteResource(ctx, customCluster, customMachine)
+		log.Info("terminating worker was completed successfully, delete the related CRD")
+		if err := r.deleteResource(ctx, customCluster, customMachine, kcp); err != nil {
+			log.Error(err, "failed to delete resource", "customCluster", customCluster.Name)
```
Comment (zirain) [inline pkg/controllers/customcluster_controller.go:365]:
```suggestion
			log.Error(err, "failed to delete resource", "name", customCluster.Name, "namespace", customCluster.Namespace)
```

File: pkg/controllers/customcluster_controller.go [hunk 5]
```diff
@@ -228,719 +237,243 @@ func (r *CustomClusterController) Reconcile(ctx context.Context, req ctrl.Reques
...
+	controllerutil.AddFinalizer(kcp, CustomClusterFinalizer)
+	controllerutil.AddFinalizer(clusterHosts, CustomClusterConfigMapFinalizer)
+	controllerutil.AddFinalizer(clusterConfig, CustomClusterConfigMapFinalizer)
 
-	ownerRefs := generateOwnerRefFromCustomCluster(res.customCluster)
-	res.customMachine.OwnerReferences = []metav1.OwnerReference{ownerRefs}
-	res.clusterHosts.OwnerReferences = []metav1.OwnerReference{ownerRefs}
-	res.clusterConfig.OwnerReferences = []metav1.OwnerReference{ownerRefs}
+	ownerRefs := generateOwnerRefFromCustomCluster(customCluster)
+	customMachine.OwnerReferences = []metav1.OwnerReference{ownerRefs}
```
Comment (zirain) [inline pkg/controllers/customcluster_controller.go:452]: use `capiutil.EnsureOwnerRef`

File: pkg/controllers/customcluster_helper.go [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"kurator.dev/kurator/pkg/apis/infra/v1alpha1"
+)
+
+// generateClusterManageWorker generate a kubespray manage worker pod from configmap.
+func (r *CustomClusterController) generateClusterManageWorker(customCluster *v1alpha1.CustomCluster, manageAction customClusterManageAction, manageCMD customClusterManageCMD, hostsName, configName string) *corev1.Pod {
```
Comment (zirain) [inline pkg/controllers/customcluster_helper.go:37]: unnecessary to be a method of `CustomClusterController`

File: pkg/controllers/customcluster_helper.go [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+	}
+	return r.CreateClusterHosts(ctx, customMachine, customCluster)
+}
+
+func (r *CustomClusterController) CreateClusterHosts(ctx context.Context, customMachine *v1alpha1.CustomMachine, customCluster *v1alpha1.CustomCluster) (*corev1.ConfigMap, error) {
+	hostsContent := GetHostsContent(customMachine)
+	hostData := &strings.Builder{}
+
+	// todo: split this to a separated file
+	tmpl := template.Must(template.New("").Parse(`
```
Comment (zirain) [inline pkg/controllers/customcluster_helper.go:203]: use std lib `embed`

File: pkg/controllers/customcluster_scale_test.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+	scaleDoneNodes4 := findScaleDownWorkerNodes(nil, curNodes2)
+	assert.Equal(t, 0, len(scaleDoneNodes4))
+
+	scaleDoneNodes5 := findScaleDownWorkerNodes(curNodes2, nil)
+	assert.Equal(t, curNodes2, scaleDoneNodes5)
+}
+
+func TestGenerateScaleDownManageCMD(t *testing.T) {
+	scaleDownCMD1 := generateScaleDownManageCMD(nodeNeedDelete1)
```
Comment (zirain) [inline pkg/controllers/customcluster_scale_test.go:60]: We typically use the following programming paradigm:
```golang
cases := []struct {
		name string
		obj  runtime.Object

		expected *policyv1alpha1.PropagationPolicy
	}{}

for _, tc := range cases {
  t.Run(tc.name, func(t *testing.T) {}
}
```

File: pkg/controllers/customcluster_controller.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/controllers/customcluster_controller.go:213]: IIRC, the controller runtime does not respect requeueAfter at all when there is an err

```
	result, err := c.Reconcile(ctx, req)
	switch {
	case err != nil:
		c.Queue.AddRateLimited(req)
		ctrlmetrics.ReconcileErrors.WithLabelValues(c.Name).Inc()
		ctrlmetrics.ReconcileTotal.WithLabelValues(c.Name, labelError).Inc()
		log.Error(err, "Reconciler error")
```

File: pkg/controllers/customcluster_scale_test.go (hunk 8 above)
Comment (Xieql) [reply pkg/controllers/customcluster_scale_test.go:60]: fix it later

File: pkg/controllers/customcluster_controller.go [hunk 9]
```diff
@@ -201,20 +195,38 @@ func (r *CustomClusterController) Reconcile(ctx context.Context, req ctrl.Reques
...
+		return ctrl.Result{}, errors.Wrapf(err1, "failed to init patch helper for cluster %s", req.NamespacedName)
 	}
 
 	defer func() {
 		if err := r.Status().Update(ctx, customCluster); err != nil {
 			log.Error(err, "failed to update customCluster status", "customCluster", req)
 			reterr = err
 		}
-
+		if err := r.Update(ctx, customCluster); err != nil {
```
Comment (hzxuzhonghu) [inline pkg/controllers/customcluster_controller.go:226]: we donot need calling updater status and update

The patch below can work

File: pkg/controllers/customcluster_controller.go [hunk 10]
```diff
@@ -228,719 +240,249 @@ func (r *CustomClusterController) Reconcile(ctx context.Context, req ctrl.Reques
...
-		if len(scaleDownWorkerNodes) != 0 || phase == v1alpha1.ScalingUpPhase {
-			return r.reconcileScaleDown(ctx, customCluster, customMachine, scaleDownWorkerNodes)
-		}
-		return ctrl.Result{}, nil
+	// Get desiredClusterInfo and provisionedClusterInfo to determine if further scaling or upgrading is needed.
+	// desiredClusterInfo contains information retrieved from configured CRDs such as "customMachine" and "kcp".
+	desiredClusterInfo := getDesiredClusterInfo(customMachine, kcp)
+	// provisionedClusterInfo contains information retrieved from configmap that represent provisioned cluster.
+	provisionedClusterInfo, err := r.getProvisionedClusterInfo(ctx, customCluster)
+	if err != nil {
```
Comment (hzxuzhonghu) [inline pkg/controllers/customcluster_controller.go:273]: shouldn't this protected by `phase == ProvisionedPhase`

File: pkg/controllers/customcluster_controller.go [hunk 11]
```diff
@@ -228,719 +240,249 @@ func (r *CustomClusterController) Reconcile(ctx context.Context, req ctrl.Reques
...
-			log.Info("terminate worker does not exist, turn to reconcileVMsTerminate to create a new one", "worker", terminateWorkerKey)
-			return r.reconcileVMsTerminate(ctx, customCluster)
-		}
-		log.Error(err, "failed to get terminate worker. maybe it has been deleted", "worker", terminateWorkerKey)
-		return ctrl.Result{RequeueAfter: RequeueAfter}, err
+	// Create the termination worker to handle the cluster deletion.
+	terminateWorker, err1 := r.ensureWorkerPodCreated(ctx, customCluster, CustomClusterTerminateAction, KubesprayTerminateCMD, generateClusterHostsName(customCluster), generateClusterConfigName(customCluster))
+	if err1 != nil {
+		conditions.MarkFalse(customCluster, v1alpha1.TerminatedCondition, v1alpha1.FailedCreateTerminateWorker,
+			clusterv1.ConditionSeverityWarning, "terminate worker not ready %s/%s.", customCluster.Namespace, customCluster.Name)
```
Comment (hzxuzhonghu) [inline pkg/controllers/customcluster_controller.go:364]: why `worker not ready`, isn't it failed to create ？

File: pkg/controllers/customcluster_helper.go [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+	configData := &strings.Builder{}
+
+	tmpl := template.Must(template.New("").Parse(clusterConfigTemplate))
+	if err := tmpl.Execute(configData, configContent); err != nil {
+		return nil, err
+	}
+	name := generateClusterConfigName(cc)
+	namespace := cc.Namespace
+
+	return r.CreateConfigMapWithTemplate(ctx, name, namespace, ClusterConfigName, configData.String())
```
Comment (hzxuzhonghu) [inline pkg/controllers/customcluster_helper.go:230]: nit: template is on above

File: pkg/controllers/customcluster_scale.go [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+	"sigs.k8s.io/cluster-api/util/conditions"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"kurator.dev/kurator/pkg/apis/infra/v1alpha1"
+)
+
+// reconcileScaleUp is responsible for handling the customCluster reconciliation process when worker nodes need to be scaled up.
+func (r *CustomClusterController) reconcileScaleUp(ctx context.Context, customCluster *v1alpha1.CustomCluster, scaleUpWorkerNodes []NodeInfo) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
```
Comment (hzxuzhonghu) [inline pkg/controllers/customcluster_scale.go:38]: With values cluster namespace/name

Comment (hzxuzhonghu) [reply pkg/controllers/customcluster_scale.go:38]: like log := ctrl.LoggerFrom(ctx).WithValues("cluster", types.NamespacedName{Name: Name, Namespace: Namespace})

File: pkg/controllers/customcluster_controller.go (hunk 11 above)
Comment (Xieql) [reply pkg/controllers/customcluster_controller.go:364]: ok， it need rename

File: pkg/controllers/customcluster_helper.go (hunk 12 above)
Comment (Xieql) [reply pkg/controllers/customcluster_helper.go:230]: ok

File: pkg/controllers/customcluster_scale.go (hunk 13 above)
Comment (Xieql) [reply pkg/controllers/customcluster_scale.go:38]: got it

File: pkg/controllers/customcluster_controller.go (hunk 9 above)
Comment (Xieql) [reply pkg/controllers/customcluster_controller.go:226]: you are right, the patch works

File: pkg/controllers/customcluster_controller.go [hunk 14]
```diff
@@ -154,13 +148,18 @@ func (r *CustomClusterController) Reconcile(ctx context.Context, req ctrl.Reques
...
-		return ctrl.Result{RequeueAfter: RequeueAfter}, err
+		return ctrl.Result{}, err
 	}
 	log = log.WithValues("customCluster", klog.KObj(customCluster))
 	ctx = ctrl.LoggerInto(ctx, log)
 	// ensure customCluster status no nil
 	if len(customCluster.Status.Phase) == 0 {
 		customCluster.Status.Phase = v1alpha1.PendingPhase
+		// using patchHelper.Patch() to update status need ensure "isConditionsSetter" is true.
+		if err := r.Status().Update(ctx, customCluster); err != nil {
```
Comment (hzxuzhonghu) [inline pkg/controllers/customcluster_controller.go:159]: what do you mean

Comment (Xieql) [reply pkg/controllers/customcluster_controller.go:159]: This is not needed, my understanding is wrong

**Review by hzxuzhonghu — APPROVED** [review]
/lgtm

**Pull request merged**